crates/red_knot_python_semantic/src/types.rs

```diff
@@ -329,19 +331,37 @@ impl<'db> FunctionType<'db> {
 #[salsa::interned]
 pub struct ClassType<'db> {
     /// Name of the class at definition
-    pub name: Name,
+    pub name: ast::name::Name,
 
-    /// Types of all class bases
-    bases: Vec<Type<'db>>,
+    definition: Definition<'db>,
 
     body_scope: ScopeId<'db>,
 }
 
 impl<'db> ClassType<'db> {
+    /// Return an iterator over the types of this class's bases.
+    ///
+    /// # Panics:
+    /// If `definition` is not a `DefinitionKind::Class`.
+    pub fn bases(&self, db: &'db dyn Db) -> impl Iterator<Item = Type<'db>> {
+        let DefinitionKind::Class(class_stmt_node) = self.definition(db).node(db) else {
+            panic!("Class type definition must have DefinitionKind::Class");
+        };
+        let definition = self.definition(db);
+        let inference = if definition.file(db).is_stub(db.upcast()) {
+            infer_deferred_types(db, definition)
+        } else {
+            infer_definition_types(db, definition)
+        };
```

Comment:
I don't love that this knowledge of whether the bases are deferred or not is spread across both type inference and here. But this explicit "look in the right place first" logic is more efficient than either an implicit "first check one, then the other" or merging/copying the two inferences. I'll think more tomorrow about whether there's a better way to consolidate this logic in one place.

Reply:
Won't we also have to support deferred bases in `.py` files? Pyright and mypy both support recursive definitions like this in `.py` files:

```py
class Foo(list["Foo | None"]): pass
```

- https://mypy-play.net/?mypy=latest&python=3.12&gist=e9e1e5ebb6a0d34f80b190a395f19f7b
- https://pyright-play.net/?pyrightVersion=1.1.368&code=GYJw9gtgBALgngBwJYDsDmUkQWEMogCmAboQIYA2A%2BvAoQFCMDGFZAzm1AGJhgAUFJGxgBtAEQ8wUAD5QAcmBSExAXQCUALigJ2bRkVKUaiQn0l81IgAzr6QA

Though I suppose we don't support stringified annotations of any kind yet. Worth a TODO comment, at least?

Reply:
Ah yeah, good call. I did have a TODO about stringified annotations (and the future import) in an earlier draft here, but that method got removed and I forgot to re-add the TODO comment.

I'm tempted to add support for deferred partially-stringified bases in `.py` files in this PR, so as to clarify how it will impact logic like this. The awkward thing is just that stringified types in bases in `.py` are only valid as generic type parameters, so I'll need to see how much of the subscripting logic needs to be added along with that, and if that starts to feel like scope creep.

In any case, I think you're right that this means we can't know in advance whether the type of a particular base is inferred or not, so I will need to provide some facility for finding an expression type that's part of a definition without knowing that in advance.

Reply:
> I'm tempted to add support for deferred partially-stringified bases in `.py` files in this PR, so as to clarify how it will impact logic like this. The awkward thing is just that stringified types in bases in `.py` are only valid as generic type parameters, so I'll need to see how much of the subscripting logic needs to be added along with that, and if that starts to feel like scope creep.

If it makes things easier -- in flake8-pyi, we only allow forward references in class bases when they're inside generic type parameters. We run flake8-pyi over the whole of typeshed with no false positives, so I think it would be fine if we decided in red-knot that we only wanted to permit forward references inside class bases if they're inside generic type parameters, even if the source file is a `.pyi` file. Since there's no known use case for forward references in class bases that are outside generic type parameters, this seems like it would make sense if it reduces complexity for our implementation.

Reply:
(I'm referring here to the [original flake8-pyi plugin for flake8](https://github.com/PyCQA/flake8-pyi), not Ruff's reimplementation of those lints. Ruff, I think, is slightly more permissive when it comes to the forward references it allows in `.pyi` files? I can't remember exactly where we landed on all of that... anyway, we're creating a semantic model "from scratch" here; there's no need to follow what Ruff does precisely.)

Reply:
Adding support for stringified generic type params really felt like it would require adding generics support. Which I'd like to do, but not in this PR. So I didn't do that. But I did make the lookup generic in such a way that `ClassType::bases` no longer makes assumptions about whether a given base is deferred or not. And I added a TODO for the stringified generic type parameters.